### PR TITLE
[Kernel][FP8] Initial support with dynamic per-tensor scaling

### PR DESCRIPTION
**Pull request by comaniac**
Provide an initial support to FP8 computation. This PR is inspired by HuggingFace TGI: https://github.com/huggingface/text-generation-inference/pull/1726

This feature can be enabled with `--quantization fp8` or `-q fp8` when launching an engine.

## Algorithm
We still load a model checkpoint in FP16/BF16. After the weights are loaded, Fp8LinearMethod calculates the per-tensor scaling factor of weights and quantizes the weights accordingly. The scaling factor will then be stored for future use. Meanwhile, the per-tensor scaling factor for activations is calculated in every forward pass.

## Initial Results

Currently tested Mistral-7B on 1xH100. With prompt length ~5 and decoding length 128:
* BF16: 1.47s
* FP8: 1.66s

I'll try to use larger models and try to find more performance bottleneck. Meanwhile, you're welcome to try this PR.

## To Do
- [ ] Unit tests
- [ ] Comprehensive benchmarking
- [ ] Refine the interface

cc @WoosukKwon @zhuohan123 @robertgshaw2-neuralmagic @mgoin 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/quantization/fp8.py
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:96]: Will this dealloc the memory allocated for the original full precision weight Parameter? Unfortunately this will not help reduce peak memory required since we will be allocating full precision weights for the entire model before quantizing down in the forward pass - but that is fine for now.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:107]: Could we add emulated support for devices without FP8 hardware? Something like this where it simply dequantizes:
```python
        cuda_compute_capability = torch.cuda.get_device_capability()
        if cuda_compute_capability >= (9, 0):
            output, _ = torch._scaled_mm(
                qinput,
                layer.weight,
                out_dtype=x.dtype,
                scale_a=scale,
                scale_b=layer.scale,
                bias=bias,
            )
        else:
            # For NVIDIA SM < 9.0
            logger.warn("FP8 hardware support doesn't exist for "
                      "NVIDIA SM < 9.0. Up-conversion to "
                      "original dtype will be used.")

            output = F.linear(
                qinput.to(x.dtype) * scale, 
                layer.weight.to(x.dtype) * layer.scale.to(x.dtype),
                bias)
```

File: vllm/model_executor/layers/quantization/fp8.py
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:24]: I don't think `torch.float8_e4m3fn, torch.float8_e5m2` will work at the moment if passed into `apply_weights` since it will always quantize. We would also need to know the `scale`, so a refactor is needed if we want to support that input

File: vllm/model_executor/layers/quantization/fp8.py
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:29]: Add a note that 89 CUDA capability should work in the future but will require a pytorch upgrade https://github.com/pytorch/pytorch/blob/72271fb07ef88c33a5cb928eec6d8af8bf365493/aten/src/ATen/native/cuda/Blas.cpp#L825

**Comment by mgoin** [comment]
Nice job here for a clean implementation. For performance improvement, I think the easiest/clearest measurement would be to measure TTFT for a large prompt since we are trying to improve compute efficiency here, rather than just memory bandwidth like most existing weight-only quantization methods in vllm.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:96]: Right this doesn't help with memory footprint at all. Like you said we have to quantize weights during weight loading, but it's not straightforward to store the scaling factor at that time.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:107]: Of course this can be done to benefit other GPUs' memory footprint. Meanwhile, the current quantization config already checks the compatibility and throws an error when running on SM <9.0 with `-q fp8`. We could think about a proper interface.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (pcmoritz) [reply vllm/model_executor/layers/quantization/fp8.py:96]: I think the right approach here is to have model files with weights that are already fp8 quantized and have the scaling factors in them -- I implemented this here https://github.com/vllm-project/vllm/pull/3954/files#diff-74473eb619768ed055303d29d74c47b1753615e8323e69467506b2b42d3f9898R139. Even if only the fully connected layers are quantized, the model files and memory use gets *a lot* smaller this way :)

Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:96]: Yes that's the easiest way as it is compatible with all existing quantization methods. Ideally I wish we could make both methods working on the other hand.

Comment (mgoin) [reply vllm/model_executor/layers/quantization/fp8.py:96]: On my first question though, once the quantization has been done at first run, will the original weight be deallocated and the memory usage decreased? I'm not sure the difference between `layer.weight.data = qweight.t()` and `layer.weight = qweight.t()`

File: vllm/model_executor/layers/quantization/fp8.py
Comment (zhuohan123) [reply vllm/model_executor/layers/quantization/fp8.py:29]: @comaniac Can you add the comment in the code as well?

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:96]: The current code is actually problematic, because the weight parameter has data type float16, but we assigned a float8 tensor, so `layer.weight = qweight.t()` will throw an error because it does a validation.

**Comment by comaniac** [comment]
I did a benchmarking with 50 requests sending with QPS 1. The average prompt length is 512 and decoding length is 128. The results are as follows:

Case | TTFT (ms) | ITL (ms)
--|--|--
FP16 | 28.8 | 8.5
FP8 | 38.8 | 10.2
FP8-test | 27.4 | 9.5

In `FP8-test`, I removed activation quantization to understand its impact. Although this results in garbage output, it shows that activation quantization does introduce a large amount of overhead. Meanwhile, even removing activation quantization, the speedup isn't obvious. This looks weird to me because in this case we only have `.to` and `torch._scaled_mm` in the forward. I'm going to perform a microbenchmark with `torch._scaled_mm` to see what speedup we could actually achieve against `F.linear`.

**Comment by mgoin** [comment]
@comaniac I did some lower-level benchmarking just on `torch._scaled_mm` as you mentioned and got some fairly interesting results that I'm still investigating. I did this on an RTX 4090  so I did have to use the pytorch-nightly build in order to support CUDA capability 8.9. You can find my code here: https://github.com/mgoin/torch-fp8

Here is a fine-grained benchmark showing various M (num tokens dim, so either batch size or prefill length) for the layer shapes of Llama 7B:

image

Obviously some interesting stuff seeing more than 2x speedup, but I wouldn't take that so seriously since this is individual layer benchmarking and it is likely fitting into cache at small M (however I am unsure about the bump at M between 100-130, very interesting)

Here is a benchmark that is less fine-grained but showing the impact of dynamic activation quantization:

image

I'll continue benchmarking and investigating how we can minimize overhead on the individual layer level and on different GPUs, but it does look like the performance of `scaled_mm` is highly variable for M<192 on an RTX 4090. This is bound to be much worse on a GPU with more memory bandwidth like an H100.

**Comment by comaniac** [comment]
Thanks for the benchmarking and it basically aligns with my results on H100. I'll also do more investigations to see whether we could amortize this overhead.

**Comment by comaniac** [comment]
@WoosukKwon @zhuohan123 @robertgshaw2-neuralmagic @mgoin @pcmoritz 

Although we haven't achieved a decent performance with this FP8 support, I'd like to get this PR reviewed first, because it seems better to build follow-up optimizations on top of this PR so merging this PR could unblock others to try more interesting stuffs (unless you have concerns about the current interface and API implementation).

In short, this PR now offers:
1. Loading FP16 checkpoints and per-tensor quantizing linear weights to FP8 during the first forward pass.
2. Per-tensor quantizing activations to FP8 in each forward pass.
3. We also accept FP8 checkpoints. It has to store linear weights in `torch.float8_e4m3fn` along with an additional `w_scale` parameter. The major benefit of loading a FP8 checkpoint is lowering the memory footprint.

Some follow-ups (I probably don't have time to do all of these so you're welcome to take anything you interest):
1. Confirm with PyTorch and NVIDIA about the performance of `torch._scaled_mm`.
2. Benchmark TGI v2.0.0 with FP8, and confirm that we are on the same boat now.
3. If the above 2 points are true, we need a high-efficient GEMM kernel for small number of tokens. @pcmoritz suggested that we could try Triton (https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html).
4. Investigate about how can we reduce/amortize activation quantization overheads. There are two possible directions. First, compute more ops in FP8 so that one quantization can benefit more. Second, leverage delayed scaling (i.e. reuse scaling factor), but this may affect the result quality significantly.
5. Enable FP8 in attention layer: Flash attention should have supported FP8 so we should be good with prefill. For decoding, we better integrate FlashInfer directly as it supports FP8 too.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (pcmoritz) [inline vllm/model_executor/layers/quantization/fp8.py:109]: Can you say why the tensor is being transposed here? Does it have something to do with `_scaled_mm` or is it also done for normal linear layers?

File: vllm/model_executor/layers/quantization/fp8.py
Comment (pcmoritz) [inline vllm/model_executor/layers/quantization/fp8.py:91]: How does the logic for loading this parameter from an FP8 checkpoint work? I thought normally it needs to be done explicitly in the `weight_loader` -- curious how it works since it would also be useful for my fp8 MoE PR :)

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:109]: This is because _scaled_mm requires column major layout.

File: vllm/entrypoints/llm.py
Comment (pcmoritz) [inline vllm/entrypoints/llm.py:45]: Let's add "(experimental)" for `fp8` here until we have better test coverage and have solved the performance problems :)

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:91]: When loading weights from a checkpoint, we iterate its tensors and assign them to the parameter with the same name. See function load_weights in llama.py for example.

**Review by pcmoritz — APPROVED** [review]
Very nicely done! We should add some tests pretty quickly as a follow up PR. I tested it locally with
```python
from vllm import LLM, SamplingParams


llm = LLM(model="mistralai/Mistral-7B-Instruct-v0.1", tensor_parallel_size=1, quantization="fp8")

prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is",
]
sampling_params = SamplingParams(temperature=0.8, top_p=0.95)

outputs = llm.generate(prompts, sampling_params)

# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
```

and making sure FP8 is indeed used with

```
In [9]: llm.llm_engine.model_executor.driver_worker.model_runner.model.model.layers[0].mlp.down_proj.linear_method
Out[9]: <vllm.model_executor.layers.quantization.fp8.Fp8LinearMethod at 0x7f66163aa280>
```

(TP2 also works, but only after master is merged into this PR -- it seems it was broken when the PR was branched off)

File: vllm/model_executor/layers/quantization/fp8.py
Comment (pcmoritz) [reply vllm/model_executor/layers/quantization/fp8.py:91]: Ah got it, the weight loading mechanism only needs to specify the mappings if parameters are renamed / reshuffled, thanks for clarifying, I wasn't aware of this :)

File: vllm/model_executor/layers/quantization/fp8.py
Comment (pcmoritz) [reply vllm/model_executor/layers/quantization/fp8.py:109]: (it wouldn't hurt to add a comment in the code that clarifies this)

**Comment by comaniac** [comment]
Thanks for the review! Will add some tests tomorrow along with your suggestions.

**Review by mgoin — COMMENTED** [review]
In order to test the FP8 weight loading, could you include an example of how you save the FP8 model format? Something like the [FP8 KV Cache example](https://github.com/vllm-project/vllm/tree/main/examples/fp8) would be nice

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:109]: I did some tests yesterday and it seems ok. When param.data is overwritten, the original tensor has no reference anymore and will be freed (torch.cuda.empty_cache may be required to reduce fragmentation).

I guess my impression comes from training, and our weights are frozen all the time in inference...

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:91]: We have no way to set it to READY because the logic of creat_weights and loading weights are separated. If FP8 weights are loaded, we simply set to READY in the first forward pass without doing anything.

File: benchmarks/benchmark_latency.py
Comment (comaniac) [reply benchmarks/benchmark_latency.py:104]: Thanks. Rebased.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:128]: Thanks for the info this is useful. Do you have any reference so that we could put it in comments?

**Comment by comaniac** [comment]
> In order to test the FP8 weight loading, could you include an example of how you save the FP8 model format? Something like the [FP8 KV Cache example](https://github.com/vllm-project/vllm/tree/main/examples/fp8) would be nice

Thanks for the pointer. I could actually make it compatible with the existing quantization flow. Meanwhile, I found that the quantizer seems incompatible with the latest AMMO. When I run quantizer.py with AMMO 0.9.4, I saw the following error:

```
Traceback (most recent call last):
  File "/home/ray/default/vllm/examples/fp8/quantizer/quantize.py", line 28, in <module>
    from ammo.torch.export import export_model_config
ImportError: cannot import name 'export_model_config' from 'ammo.torch.export' (/home/ray/anaconda3/lib/python3.10/site-packages/ammo/torch/export/__init__.py)
```

We probably need to update the script accordingly, as AMMO 0.7.1 depends on PyTorch 1.11 which is also out of date.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (mgoin) [reply vllm/model_executor/layers/quantization/fp8.py:128]: Yes please reference https://github.com/pytorch/pytorch/blob/2e48b39603411a41c5025efbe52f89560b827825/aten/src/ATen/native/cuda/Blas.cpp#L854-L856

**Comment by comaniac** [comment]
After more investigations, it seems not straightforward to support both FP16 and FP8 checkpoints in the same Linear implementation, so this PR would focus on FP16 checkpoints first (I'll explain the reasons in details but that's tl;dr). We will probably have an RFC discussing FP8 checkpoint standards later.

---
**[tl; dr] Issues with FP8 checkpoint**

The FP8 checkpoint should include
1. weights in `torch.float8_e4m3fn`.
2. The scaling factor of each weight parameter.

The major issue is most HF checkpoints save `proj_q`, `proj_k`, `proj_v`, `gate_proj`, `up_proj` to separate tensors; while we always merge them to `qkv` and `gate_up_proj`. Thus, how to merge their scaling factors becomes an issue, because the scaling factor should be based on the input tensor of `torch._scaled_mm` (and all future supported ops). There are two cases based on different quantization methods:

1. **The FP8 checkpoint was quantized without calibration data.** In this case, when we are loading q, k, and v incrementally, we dequantize partial loaded qkv tensor back to float16 using the current scaling factor, add the new loaded tensor, and re-quantize them together to get the new float8 tensor as well as the updated scaling factor. While this approach is doable, it requires some changes in the interface, because the current `weight_loader` is completely isolated from other parameters.
2. **The FP8 checkpoint was quantized with calibration data (e.g., AMMO, SmoothQuant, etc).** In this case, we cannot do anything unfortunately, unless we adopt per-channel scaling factor.

I guess supporting both cases is inevitable, so we should definitely focus more on (2) as its solution must cover (1). Just for reference, TensorRT-LLM solves this by first loading the weight to their model implementation (so qkv is already fused), performs quantization with calibration data, and save the FP8 checkpoint for their model (so the parameter names are diverse from the original HF model). We can probably do the same, but that means every model wants to run on vLLM with FP8 has to use our quantization flow. Anyways, we could have more discussions later.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (comaniac) [reply vllm/model_executor/layers/quantization/fp8.py:89]: IIUC this is only registered to XXLinear modules when the linear method is Fp8LinearMethod? Please let me know if I missed anything.

File: vllm/model_executor/layers/quantization/fp8.py
Comment (pcmoritz) [inline vllm/model_executor/layers/quantization/fp8.py:122]: This should be removed

**Pull request merged**
